File: apps/zxtune-android/zxtune/src/main/java/app/zxtune/TimeStamp.java
/**
 * @file
 * @brief Time stamp value type
 * @author vitamin.caig@gmail.com
 */

package app.zxtune;

import java.util.Locale;
import java.util.concurrent.TimeUnit;

public final class TimeStamp implements Comparable<TimeStamp> {

  private static final TimeUnit UNIT = TimeUnit.MILLISECONDS;
  private final long value;

  private TimeStamp(long val, TimeUnit unit) {
    this.value = UNIT.convert(val, unit);
  }

  public static final TimeStamp EMPTY = new TimeStamp(0, UNIT);

  public static TimeStamp fromMicroseconds(long val) {
    return new TimeStamp(val, TimeUnit.MICROSECONDS);
  }

  public static TimeStamp fromMilliseconds(long val) {
    return new TimeStamp(val, TimeUnit.MILLISECONDS);
  }

  public static TimeStamp fromSeconds(long val) {
    return new TimeStamp(val, TimeUnit.SECONDS);
  }

  public static TimeStamp fromDays(int val) {
    return new TimeStamp(val, TimeUnit.DAYS);
  }

  public final long toMicroseconds() {
    return UNIT.toMicros(value);
  }

  public final long toMilliseconds() {
    return UNIT.toMillis(value);
  }

  public final long toSeconds() {
    return UNIT.toSeconds(value);
  }

  @Override
  public String toString() {
    final int totalSec = (int) UNIT.toSeconds(value);
    final int totalMin = totalSec / 60;
    final int totalHour = totalMin / 60;
    final int min = totalMin % 60;
    final int sec = totalSec % 60;
    return totalHour != 0 ? String.format(Locale.US, "%d:%02d:%02d", totalHour, min, sec) : String
        .format(Locale.US, "%d:%02d", min, sec);
  }

  @Override
  public int hashCode() {
    return (int) (value ^ (value >>> 32));
  }

  @Override
  public boolean equals(Object rh) {
    return this == rh || (rh != null && rh.getClass() == getClass() && value == ((TimeStamp) rh).value);
  }

  @Override
  public int compareTo(TimeStamp rh) {
    return Long.compare(value, rh.value);
  }

  public final TimeStamp multiplies(long count) {
    return new TimeStamp(value * count, UNIT);
  }
}

File: apps/zxtune-android/zxtune/src/main/java/app/zxtune/TimeStamp.kt
/**
 * @file
 * @brief Time stamp value type
 * @author vitamin.caig@gmail.com
 */
package app.zxtune

import java.util.concurrent.TimeUnit

private val UNIT = TimeUnit.MILLISECONDS

class TimeStamp private constructor(private val value: Long) : Comparable<TimeStamp> {

    private constructor(input: Long, unit: TimeUnit) : this(UNIT.convert(input, unit))

    fun toMicroseconds() = UNIT.toMicros(value)

    fun toMilliseconds() = UNIT.toMillis(value)

    fun toSeconds() = UNIT.toSeconds(value)

    override fun toString() = UNIT.toSeconds(value).let { totalSec ->
        val totalMin = totalSec / 60
        val totalHour = totalMin / 60
        val min = totalMin % 60
        val sec = totalSec % 60
        if (totalHour != 0L)
            "%d:%02d:%02d".format(totalHour, min, sec)
        else
            "%d:%02d".format(min, sec)
    }

    override fun hashCode() = (value xor (value ushr 32)).toInt()

    override fun equals(other: Any?) = value == (other as? TimeStamp)?.value

    override operator fun compareTo(other: TimeStamp) = when {
        value == other.value -> 0
        value < other.value -> -1
        else -> +1
    }

    fun multiplies(count: Long) = TimeStamp(value * count)

    companion object {
        @JvmField
        val EMPTY = TimeStamp(0, UNIT)

        @JvmStatic
        fun fromMicroseconds(value: Long) = TimeStamp(value, TimeUnit.MICROSECONDS)

        @JvmStatic
        fun fromMilliseconds(value: Long) = TimeStamp(value, TimeUnit.MILLISECONDS)

        @JvmStatic
        fun fromSeconds(value: Long) = TimeStamp(value, TimeUnit.SECONDS)

        @JvmStatic
        fun fromDays(value: Long) = TimeStamp(value, TimeUnit.DAYS)
    }
}
